docs/msi_with_credential_design.md

```diff
@@ -0,0 +1,233 @@
+- ESTS validates the credential and issues an **access token**.
+- The access token is then used to authenticate with Azure services.
+
+## Certificate Handling
+
+To start the flow, MSAL requires a certificate. MSAL follows these steps:
+
+1. **Check for an existing certificate**: MSAL looks for a specific certificate (`devicecert.mtlsauth.local`).
+2. **Create a new certificate if not found**: If the expected certificate is not available, MSAL generates one dynamically for authentication.
+
+## Source Detection Logic
+
+MSAL follows a source detection process to determine how to interact with MSI endpoints and acquire tokens.
+
+### Environment Variable Check
+
+MSAL checks for Azure resource type based on specific environment variables to determine if the application is running on:
+
+- **Service Fabric**
+- **App Service**
+- **Azure Arc**
+- **Cloud Shell**
+- **Machine Learning**
+
+If identified, MSAL will use the appropriate legacy MSI endpoint for that resource.
+
+### Fallback to IMDS
+
+- If no specific Azure resource is identified from the environment variables, MSAL will fall back to IMDS (VMs and VMSS).
+- In this new design, Before fully falling back to IMDS, MSAL will now **probe the Credential Endpoint in IMDS**.
+- MSAL probes to see if the `/credential` endpoint exists in IMDS.
+- If the `/credential` endpoint is unavailable, it falls back to the legacy `/token` endpoint.
+
+## MSI V2 /credential Endpoint Details
+
+### Short-Lived Credential Retrieval
+
+- The `/credential` endpoint provides a **temporary credential** instead of a direct access token.
+- This credential is only valid for a short duration (1 hour) and must be used **immediately** to acquire an access token from ESTS.
+- This mechanism improves security by reducing the lifetime of sensitive authentication materials.
+
+## Steps for MSI V2 Authentication
+
+This section outlines the necessary steps to acquire an access token using the MSI V2 `/credential` endpoint. 
+
+### 1. Check for an Existing (Platform) Certificate
+- Search for a specific certificate (`devicecert.mtlsauth.local`) in `Cert:\LocalMachine\My`.
+- If found, extract its thumbprint and use it for authentication.
+
+### 2. Generate a New Certificate (if specific certificate is not found)
+- Create a new self-signed certificate with a 90-day validity.
+- Ensure the certificate has:
+  - Subject name `CN=mtls-auth` (name not final).
+  - Exportable key policy.
+
+### 3. Extract Certificate Data
+- Convert the certificate to a Base64-encoded string (`x5c`).
+- Format the JSON payload containing the certificate details for request authentication.
+
+### 4. Request MSI Credential
+- Send a POST request to the IMDS `/credential` endpoint with the certificate details.
+- The request must include:
+  - `Metadata: true` header.
+  - `X-ms-Client-Request-id` header with a GUID.
+  - JSON body containing the certificate's public key in `jwk` format.
```

Comment:
Do you want to reference the RFC? https://datatracker.ietf.org/doc/html/rfc7517#appendix-A.1